gql_error_link/lib/gql_error_link.dart

```diff
@@ -0,0 +1,68 @@
+/// GQL Link to handle execution errors and exceptions
+library gql_error_link;
+
+import "dart:async";
+import "package:gql_link/gql_link.dart";
+import "package:gql_exec/gql_exec.dart";
+
+/// A handler of GraphQL errors.
+typedef ErrorHandler = Stream<Response> Function(
+  Request request,
+  NextLink forward,
+  Response response,
+);
+
+/// A handler of Link Exceptions.
+typedef ExceptionHandler = Stream<Response> Function(
+  Request request,
+  NextLink forward,
+  LinkException exception,
+);
+
+/// [ErrorLink] allows interception of GraphQL errors (using [onError])
+/// and [LinkException]s (using [onException]).
+///
+/// In both cases [ErrorLink] transfers control over to the handler which is
+/// then responsible for yielding [Response]s.
+class ErrorLink extends Link {
+  final ErrorHandler onError;
+  final ExceptionHandler onException;
+
+  ErrorLink({
+    this.onError,
+    this.onException,
+  });
+
+  @override
+  Stream<Response> request(Request request, [forward]) async* {
```

Comment:
I believe async generators in Dart terminate after receiving an error (which is why I opted for a StreamTransformer). I presume we want ErrorLink to be recoverable, especially when using a WebSocketLink as the terminating link.

Reply:
You are most likely right about it!
I should write tests to verify that behavior.

I'll try [Result](https://pub.dev/documentation/async/latest/async/Result-class.html) stream. To me async generators are easier to write, read and understand (with the exception of the error behavior). If it works out well, I'm considering a breaking change in Links from `Stream<Response>` to `Stream<Result<Response>>`.

Reply:
Just to make sure I understand, are you suggesting that all errors would be passed in the Result, meaning that we wouldn't have to worry about code in the async generator throwing? 

I agree that async generators are easier to read and understand. However, I'm concerned about introducing an alternative mechanism for handling errors when Dart Streams already support both data and error events. Users implementing custom links wouldn't be able to rely on the built-in stream error handling.

Reply:
> Just to make sure I understand, are you suggesting that all errors would be passed in the Result, meaning that we wouldn't have to worry about code in the async generator throwing?

Yes.

>  However, I'm concerned about introducing an alternative mechanism for handling errors when Dart Streams already support both data and error events.

I understand your concerns. `Result.captureStream` and `Result.releaseStream` allow easy transition between the std streams and `Result` streams.
`Result` from `package:async` which is pretty much the extended std lib of Dart. 

Anyway, I'm now writing tests so that I can compare `Result` stream vs std stream.

Reply:
Take a look at this: https://github.com/gql-dart/gql/pull/103/commits/441823ff23343fcdc4fdc0557c46ac8341299921

I worked around the problem by using a `Result` stream. If `Result` was part of the Link API, the capture/release bookkeeping would not be necessary.

Reply:
@smkhalsa, I've implemented more tests. Do you think there's a better way to write that link?

Reply:
I finally stumbled upon a simple workaround to the generator problem:
```
  final generate = () async* {
    yield 1;
    yield* Stream<int>.error("a");
    yield* Stream<int>.error("b");
    yield 2;
  };
```

Reply:
Apologies for the noise, I worked way too late to realize the edge cases.

I played around and realized that changing the Link API is not worth it as it becomes too complicated in other use cases.

I think the current approach where `Result` is used internally is the best way to go about it.

Also I aligned the implementation to `apollo-link-error`. So it's no longer possible to both return a new stream and continue the original.